File: app/src/main/java/ru/neosvet/vestnewage/loader/page/PageLoader.java
package ru.neosvet.vestnewage.loader.page;

import android.content.ContentValues;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;

import ru.neosvet.vestnewage.data.DataBase;
import ru.neosvet.vestnewage.data.DateUnit;
import ru.neosvet.vestnewage.loader.basic.Loader;
import ru.neosvet.vestnewage.network.NeoClient;
import ru.neosvet.vestnewage.network.NetConst;
import ru.neosvet.vestnewage.storage.PageStorage;
import ru.neosvet.vestnewage.utils.Const;
import ru.neosvet.vestnewage.utils.Lib;

public class PageLoader implements Loader {
    private final NeoClient client;
    private int k_requests = 0;
    private long time_requests = 0;
    private final PageStorage storage = new PageStorage();
    private boolean isFinish = true;

    public PageLoader(NeoClient client) {
        this.client = client;
    }

    public boolean isFinish() {
        return isFinish;
    }

    public boolean download(String link, boolean singlePage) throws Exception {
        isFinish = false;
        // если singlePage=true, значит страницу страницу перезагружаем, а счетчики обрабатываем
        storage.open(link, true);
        if (!singlePage && storage.existsPage(link)) {
            return false;
        }
        if (storage.isDoctrine()) {
            downloadDoctrinePage(link);
            if (singlePage)
                storage.close();
            return true;
        }
        if (!singlePage)
            checkRequests();
        String s = link;
        int k = 1;
        if (link.contains("#")) {
            k = Integer.parseInt(s.substring(s.indexOf("#") + 1));
            s = s.substring(0, s.indexOf("#"));
            if (link.contains("?")) s += link.substring(link.indexOf("?"));
        }

        int n = k;
        boolean boolArticle = storage.isArticle();
        PageParser page = new PageParser(client);
        page.load(NetConst.SITE + Const.PRINT + s, "");
        if (singlePage)
            storage.deleteParagraphs(storage.getPageId(link));

        ContentValues row;
        int id = 0, bid = 0;
        s = page.getCurrentElem();
        do {
            if (page.isHead()) {
                k--;
                if (k == -1 && !boolArticle) {
                    n++;
                    if (link.contains("#"))
                        link = link.substring(0, link.indexOf("#"));
                    link += "#" + n;
                    k = 0;
                }
                if (k == 0) {
                    id = storage.getPageId(link);
                    row = new ContentValues();
                    row.put(Const.TIME, System.currentTimeMillis());

                    if (id == -1) { // id не найден, материала нет - добавляем
                        if (link.contains("#")) {
                            id = bid;
                            row = new ContentValues();
                            row.put(DataBase.ID, id);
                            row.put(DataBase.PARAGRAPH, s);
                            storage.insertParagraph(row);
                        } else {
                            row.put(Const.TITLE, getTitle(s, storage.getName()));
                            row.put(Const.LINK, link);
                            id = (int) storage.insertTitle(row);
                            //обновляем дату изменения списка:
                            row = new ContentValues();
                            row.put(Const.TIME, System.currentTimeMillis());
                            storage.updateTitle(1, row);
                        }
                    } else { // id найден, значит материал есть
                        //обновляем заголовок
                        row.put(Const.TITLE, getTitle(s, storage.getName()));
                        //обновляем дату загрузки материала
                        storage.updateTitle(id, row);
                        //удаляем содержимое материала
                        storage.deleteParagraphs(id);
                    }
                    bid = id;
                    s = page.getNextElem();
                }
            }
            if ((k == 0 || boolArticle) && !isEmpty(s)) {
                row = new ContentValues();
                row.put(DataBase.ID, id);
                row.put(DataBase.PARAGRAPH, s);
                storage.insertParagraph(row);
            }
            s = page.getNextElem();
        } while (s != null);

        if (singlePage)
            finish();
        page.clear();
        return true;
    }

    private void downloadDoctrinePage(String link) throws Exception {
        String s = link.substring(Const.DOCTRINE.length()); //pages
        BufferedInputStream stream = client.getStream(NetConst.DOCTRINE_BASE + s + ".p");
        BufferedReader br = new BufferedReader(new InputStreamReader(stream, Const.ENCODING), 1000);
        long time = Long.parseLong(br.readLine());
        ContentValues row = new ContentValues();
        row.put(Const.TIME, time);
        storage.updateTitle(link, row);
        int id = storage.getPageId(link);
        storage.deleteParagraphs(id);
        s = br.readLine();
        while (s != null) {
            row = new ContentValues();
            row.put(DataBase.ID, id);
            row.put(DataBase.PARAGRAPH, s);
            storage.insertParagraph(row);
            s = br.readLine();
        }
        br.close();
    }

    private boolean isEmpty(String s) {
        return Lib.withOutTags(s).isEmpty();
    }

    private void checkRequests() {
        k_requests++;
        if (k_requests == 5) {
            long now = System.currentTimeMillis();
            k_requests = 0;
            if (now - time_requests < DateUnit.SEC_IN_MILLS) {
                try {
                    Thread.sleep(400);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            time_requests = now;
        }
    }

    private String getTitle(String line, String name) {
        line = Lib.withOutTags(line).replace(".20", ".");
        if (line.contains(name)) {
            line = line.substring(9);
            if (line.contains(Const.KV_OPEN))
                line = line.substring(line.indexOf(Const.KV_OPEN) + 1, line.length() - 1);
        }
        return line;
    }

    public void finish() {
        isFinish = true;
        storage.close();
    }

    @Override
    public void cancel() {
        finish();
    }
}

File: app/src/main/java/ru/neosvet/vestnewage/loader/page/PageLoader.kt
package ru.neosvet.vestnewage.loader.page

import ru.neosvet.vestnewage.App
import ru.neosvet.vestnewage.R
import ru.neosvet.vestnewage.data.DateUnit
import ru.neosvet.vestnewage.loader.BookLoader
import ru.neosvet.vestnewage.loader.basic.Loader
import ru.neosvet.vestnewage.network.NeoClient
import ru.neosvet.vestnewage.network.Urls
import ru.neosvet.vestnewage.storage.PageStorage
import ru.neosvet.vestnewage.utils.Const
import ru.neosvet.vestnewage.utils.fromHTML
import java.io.BufferedReader
import java.io.InputStreamReader

class PageLoader(private val client: NeoClient) : Loader {
    private var kRequests = 0
    private var timeRequests: Long = 0
    private val storage = PageStorage()
    var isFinish = true
        private set

    override fun load() {
    }

    override fun cancel() {
        finish()
    }

    fun download(link: String, singlePage: Boolean): Boolean {
        isFinish = false
        // если singlePage=true, значит страницу страницу перезагружаем, а счетчики обрабатываем
        storage.open(link, true)
        val exists = storage.existsPage(link)
        if (!singlePage && exists)
            return false
        if (storage.isOtherBook) {
            downloadOtherPage(link)
            if (singlePage) storage.close()
            return true
        }
        if (!singlePage) checkRequests()
        val page = PageParser(client)
        page.load(Urls.Page + link, "")
        if (singlePage) storage.deleteParagraphs(storage.getPageId(link))
        val fix = link.contains("02.02.20") || link.contains("16.01.19") ||
                link.contains("08.10.18")

        var par = ""
        var hasNoind = false
        var s: String? = page.currentElem
        val time = System.currentTimeMillis()
        var id = link.indexOf("#")
        val adr = if (id > 0) {
            par = link.substring(id)
            link.substring(0, id)
        } else link
        id = 0
        do {
            page.link?.let {
                if (it.startsWith("name:")) {
                    par = "#" + it.substring(5)
                    s = page.nextElem
                }
            }
            if (page.isHead) {
                if (fix && s?.contains("02.02.20") == true)
                    s = s?.replace("20.20", "2020")
                hasNoind = false
                s = getTitle(s, storage.name)
                if (id > 0) storage.insertParagraph(
                    id, "<p class='noind'>" + App.context.getString(R.string.on_same_day) +
                            "<br><a href='${adr + par}'>${s?.replace("“", "")}</a></p>"
                )
                id = storage.putTitle(s!!, adr + par, time)
                if (exists) storage.deleteParagraphs(id)
                s = page.nextElem
            }
            s?.let {
                var e = if (it.contains("<br>"))
                    it.replace(" <br>", "<br>").replace("<br> ", "<br>")
                else it
                if (!hasNoind || e.contains("noind")) {
                    if (e.fromHTML.isNotEmpty()) {
                        if (e.contains("Аминь") && !e.contains("noind")) {
                            e = "<p class='noind'>" + e.substring(e.indexOf('>') + 1)
                            page.nextElem?.let {
                                e = e.substring(0, e.lastIndexOf("</"))
                                e += "<br>" + it.substring(it.indexOf('>') + 1)
                            }
                        }
                        storage.insertParagraph(id, e)
                    }
                }
                if (fix && hasNoind) {
                    hasNoind = false
                    s = if (e.startsWith("<p><br>")) {
                        //<p><br>Дополнение к Катрену через 5 часов от Создателя:</p> (16.01.19)
                        "<p class='noind'><strong" + e.substring(6, e.length - 2) + "strong></p>"
                    } else if (e.contains(":<"))//<p>Дополнение:</p> (08.10.18)
                        "<p class='noind'" + e.substring(2)
                    else null
                    s?.let { storage.insertParagraph(id, it) }
                }
                if (fix && e.contains("noind"))
                    hasNoind = true //for ignore Дополнение Информации от Создателя (02.02.20)
            }
            s = page.nextElem
        } while (s != null)
        if (singlePage) finish()
        page.clear()
        return true
    }

    private fun downloadOtherPage(link: String) {
        val isRus = link.contains(Const.HOLY_RUS)
        var id = storage.getPageId(link)
        if (id == -1) {
            val loader = BookLoader(client)
            loader.loadBookList(isRus)
            id = storage.getPageId(link)
        } else storage.deleteParagraphs(id)

        val m = if (isRus) arrayOf(Urls.HolyRusBase, Const.HOLY_RUS)
        else arrayOf(Urls.DoctrineBase, Const.DOCTRINE)
        var s: String? = link.substring(m[1].length) //pages
        val stream = client.getStream("${m[0]}$s.txt")
        val br = BufferedReader(InputStreamReader(stream, Const.ENCODING), 1000)
        storage.updateTime(link, br.readLine().toLong())
        s = br.readLine()
        while (s != null) {
            storage.insertParagraph(id, s)
            s = br.readLine()
        }
        br.close()
    }

    private fun checkRequests() {
        kRequests++
        if (kRequests == 5) {
            val now = System.currentTimeMillis()
            kRequests = 0
            if (now - timeRequests < DateUnit.SEC_IN_MILLS) {
                try {
                    Thread.sleep(400)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            timeRequests = now
        }
    }

    private fun getTitle(line: String?, name: String): String {
        if (line == null) return ""
        var s = line.fromHTML.replace(".20", ".")
        if (s.contains(name)) {
            s = s.substring(9).replace(Const.KV_CLOSE, "")
            if (s.contains("№"))
                s = s.substring(s.indexOf("№"))
            else if (s.contains(Const.KV_OPEN))
                s = s.substring(s.indexOf(Const.KV_OPEN) + 1)
        }
        return s
    }

    fun finish() {
        isFinish = true
        storage.close()
    }
}